client/v3/kubernetes/client.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2024 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"context"
+
+	pb "go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+// New creates Client from config.
+// Caller is responsible to call Close() to clean up client.
+func New(cfg clientv3.Config) (*Client, error) {
+	c, err := clientv3.New(cfg)
+	if err != nil {
+		return nil, err
+	}
+	kc := &Client{
+		Client: c,
+		kv:     clientv3.RetryKVClient(c),
+	}
+	kc.Kubernetes = kc
+	return kc, nil
+}
+
+type Client struct {
+	*clientv3.Client
+	Kubernetes Interface
```

Comment:
why add this?
```suggestion
```

Reply:
Having an explicit interface is a feature of etcd client that Kubernetes uses to inject it's own hooks. For example request latency tracking.

Reply:
See https://github.com/kubernetes/kubernetes/pull/125258/commits/0eb566c2ee115d025d4d931b9cc68be5caa459bd

Reply:
Added a comment https://github.com/kubernetes/kubernetes/pull/125258#discussion_r1669135958

It isn't necessary to add such an additional indirect reference, because `kubernetes.Client` always implements `kubernetes.Interface`.

Actually an even better implementation is to rename `Get` to a different name (e.g. `KGet`) so as to be different from the `kv.Get`,https://github.com/etcd-io/etcd/blob/24e05998c68f481af2bd567ee104cf89879bbf89/client/v3/kv.go#L49

Then you don't need the `clientV3KVLatencyTracker` at all. You can use the `*kubernetes.Client` (returned by `kubernetes.New`) in either places directly. See also https://github.com/etcd-io/etcd/pull/16333#discussion_r1665819091

Reply:
> Then you don't need the `clientV3KVLatencyTracker` at all. You can use the `*kubernetes.Client` (returned by `kubernetes.New`) in either places directly. See also [#16333 (comment)](https://github.com/etcd-io/etcd/pull/16333#discussion_r1665819091)

This might not be accurate, but the high level idea is that you don't need to differentiate an etcd client or a kubernete client,  a *kubernetes.Client (returned by kubernetes.New) can be used for both cases directly.

Reply:
Not sure I understand, the Kubernetes currently wraps only `KV` interface that is part of client, not a full *clientv3.Client. With introduction of interface I plan to wrap `Kubernetes` interface and not the whole `kubernetes.Client`.